### PR TITLE
Combine primitives

### PR DESCRIPTION
**Pull request by lasalvavida**
This is worth an initial look and some testing while I add tests for everything.

This pull request contains a few things:
1. Re-adds combine primitives, closing out #108 and by proxy #85 and #103.
2. Splits out some primitive helper functions from combineNodes into PrimitiveHelpers.
3. Removes underscore usage #160.
3. mergeBuffers will no longer fail if the new buffer name is the same as an already existing one.
4. Fixes an error in `removeDuplicatePrimitives.removePrimitivesFromMesh`
5. Adds a second pass of light optimization

Some preliminary testing with the buggy model:

|  | Original | Master | Master w/ combine primitives |
| --- | --- | --- | --- |
| File size | 10,414 KB | 6,091 KB | 6,046 KB |
| Nodes | 208 | 179 | 179 |
| Meshes | 109 | 163 | 68 |
| Draw Calls | 236 | 273 | 236 |

### CODE REVIEW

File: lib/PrimitiveHelpers.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+var Cesium = require('cesium');
+var deepEqual = require('deep-equal');
+
+var Cartesian3 = Cesium.Cartesian3;
+var Matrix4 = Cesium.Matrix4;
+var defined = Cesium.defined;
+
+var AccessorReader = require('./AccessorReader');
+var readAccessor = require('./readAccessor');
+var writeAccessor = require('./writeAccessor');
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: Same note about include order from the other PR.

File: lib/PrimitiveHelpers.js [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    getAllPrimitives : getAllPrimitives,
+    getPrimitivesByMaterialMode : getPrimitivesByMaterialMode,
+    markPrimitiveConflicts : markPrimitiveConflicts,
+    primitiveEquals : primitiveEquals,
+    primitivesShareAttributeAccessor : primitivesShareAttributeAccessor,
+    primitivesHaveOverlappingIndexAccessors : primitivesHaveOverlappingIndexAccessors,
+    transformPrimitives : transformPrimitives
+};
+
+
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: Spacing

File: lib/combinePrimitives.js [hunk 3]
```diff
@@ -1,255 +1,310 @@
...
 
-var byteLengthForComponentType = require('./byteLengthForComponentType');
-var getAccessorByteStride = require('./getAccessorByteStride');
+var PrimitiveHelpers = require('./PrimitiveHelpers');
+var findAccessorMinMax = require('./findAccessorMinMax');
+var getUniqueId = require('./getUniqueId');
 var numberOfComponentsForType = require('./numberOfComponentsForType');
-var readBufferComponentType = require('./readBufferComponentType');
-var writeBufferComponentType = require('./writeBufferComponentType');
+var readAccessor = require('./readAccessor');
```
Comment (lilleyse) [inline lib/combinePrimitives.js]: Include order

File: lib/combinePrimitives.js [hunk 4]
```diff
@@ -1,255 +1,310 @@
...
-}
+/**
+ * Combines all of the provided primitives if possible.
+ * If primitives have no shared data with the other primitives, their attributes and indices will be concatenated together and
+ * the indices will be incremented by the attribute offset. If the primitives have the same attribute accessors but different index
+ * accessors, the index accessors will be concatenated. In all other cases, the primitives will be left alone.
+ *
+ * @param gltf
+ * @param primitives
+ * @returns {*}
```
Comment (lilleyse) [inline lib/combinePrimitives.js]: This doc is incomplete.

File: lib/combinePrimitives.js [hunk 5]
```diff
@@ -1,255 +1,310 @@
...
-        "type": referenceAccessor.type,
-        "extras": {
-            "_pipeline": {
-                "deleteExtras": true
+    gltf.buffers[bufferId] = {
+        byteLength : attributeBuffer.length,
+        type : 'arraybuffer',
+        extras : {
+            _pipeline : {
+                source : attributeBuffer
```
Comment (lilleyse) [inline lib/combinePrimitives.js]: I think this should have additional properties like `mergeBuffers`
```javascript
source: source,
extension: ".bin",
deleteExtras: true
```

File: lib/combinePrimitives.js [hunk 6]
```diff
@@ -1,255 +1,310 @@
...
-                writeBufferComponentType(accessorSource, accessor.componentType, value, accessorSourceOffset + k * componentByteLength);
-                min[k] = Math.min(min[k], value);
-                max[k] = Math.max(max[k], value);
+    };
+    gltf.buffers[bufferId] = {
+        byteLength : indicesBuffer.length,
+        type : 'arraybuffer',
+        extras : {
+            _pipeline : {
+                source : indicesBuffer
```
Comment (lilleyse) [inline lib/combinePrimitives.js]: Same here.

File: lib/gltfPipeline.js [hunk 7]
```diff
@@ -79,21 +79,31 @@ function processJSONWithExtras(gltfWithExtras, options) {
...
+    removeDuplicatePrimitives(gltfWithExtras);
+    combinePrimitives(gltfWithExtras);
+    combineMeshes(gltfWithExtras);
+    removeUnused(gltfWithExtras, stats);
     generateNormals(gltfWithExtras);
-    // TODO: Combine primitives can be uncommented and added back into the pipeline once it is fixed, but right now there are too many issues with it to allow in the main pipeline.
-    // combinePrimitives(gltfWithExtras);
-    // Merging duplicate vertices again to prevent repeat data in newly combined primitives
-    // mergeDuplicateVertices(gltfWithExtras);
+    mergeDuplicateVertices(gltfWithExtras);
```
Comment (lilleyse) [inline lib/gltfPipeline.js]: As we talked about offline it may be worth testing if we still need this here, or if the final `mergeDuplicateVertices` is good enough.

File: lib/readAccessor.js [hunk 8]
```diff
@@ -20,9 +20,11 @@ module.exports = readAccessor;
  * @param {Object} gltf
  * @param {Object} accessor
  * @param {Array} results
+ * @param {Boolean} [generateAttributes=false]
```
Comment (lilleyse) [inline lib/readAccessor.js]: Should this be `[generateAttributes=true]`?

File: lib/removeDuplicatePrimitives.js [hunk 9]
```diff
@@ -1,12 +1,12 @@
 'use strict';
 var Cesium = require('cesium');
-var _ = require('underscore');
+
+var defined = Cesium.defined;
 
 var NodeHelpers = require('./NodeHelpers');
+var PrimitiveHelpers = require('./PrimitiveHelpers');
 var getUniqueId = require('./getUniqueId');
```
Comment (lilleyse) [inline lib/removeDuplicatePrimitives.js]: Fix require order here too. (And alphabetical).

File: package.json [hunk 10]
```diff
@@ -33,6 +33,7 @@
     "clone": "1.0.2",
     "data-uri-to-buffer": "0.0.4",
     "datauri": "1.0.4",
+    "deep-equal": "^1.0.1",
```
Comment (lilleyse) [inline package.json]: Remove the ^

**Comment by lilleyse** [comment]
Looks great! I did some simple testing and it seems to work well.

**Comment by pjcozzi** [comment]
@lasalvavida can you verify that it makes sense for the buggy model to have 95 fewer meshes, but only 37 fewer draw calls?

File: lib/PrimitiveHelpers.js [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+var Cartesian3 = Cesium.Cartesian3;
+var Matrix4 = Cesium.Matrix4;
+var defined = Cesium.defined;
+
+var AccessorReader = require('./AccessorReader');
+var readAccessor = require('./readAccessor');
+var writeAccessor = require('./writeAccessor');
+
+module.exports = {
```
Comment (pjcozzi) [inline lib/PrimitiveHelpers.js]: @lilleyse @lasalvavida can you two please review the design of this class?  I don't think it is needed.  Some functions could be private in the files they are used in; others could be standalone functions.

Let me know if you need help.

**Comment by lasalvavida** [comment]
@pjcozzi Master currently bloats the number of meshes in cases where two meshes that are in node endpoints that can't be combined contain identical primitives. That is what drives up the draw call count. The output with this PR removes that bloating by recombining meshes after primitives get combined.

tl;dr It has the same number of draw calls as the version in the sampleModels, the current output of master is adding more unnecessary draw calls.

**Comment by pjcozzi** [comment]
Gotcha.

**Comment by lilleyse** [comment]
@lasalvavida Is there more to do for this? The latest merge commit has a few issues.

**Comment by lasalvavida** [comment]
There was a bad merge; it still needs tests. I'll add those now.

**Comment by lasalvavida** [comment]
@lilleyse This is updated and ready for a look

**Comment by lasalvavida** [comment]
@lilleyse, I've done some cleanup here to get this in line with master again, and I've pulled the `createAccessor` utility from the hard-normals branch to simplify the code a bit. I've tested a few models, and everything looks good on my end. Let me know if there's anything else that needs to be done here before merging.

**Comment by lilleyse** [comment]
I'll finish reviewing this soon.

Not super relevant to this PR, but does the second `mergeDuplicateVertices` call in `Pipeline.processJSONWithExtras` have much effect anymore? It would be great if we could remove that one if possible to help improve performance.

**Comment by lasalvavida** [comment]
I have removed the second `mergeDuplicateVertices` call. With the light optimization, it is no longer needed; the files produced with and without it are the same size.

**Comment by pjcozzi** [comment]
@lilleyse will you be able to review this soon?  We have a lot of open PRs in this branch and I would like to get them merged.

**Comment by lilleyse** [comment]
Yes I will review this today.

File: lib/Pipeline.js [hunk 12]
```diff
@@ -105,8 +107,17 @@ Pipeline.processJSONWithExtras  = function(gltfWithExtras, options) {
...
-    // Run removeAll stage again after all pipeline stages have been run to remove objects that become unused
+    // Run through light optimization for newly created data
+    MergeDuplicateProperties.mergeAccessors(gltfWithExtras);
+    removeDuplicatePrimitives(gltfWithExtras);
+    combineNodes(gltfWithExtras);
+    combineMeshes(gltfWithExtras);
+    MergeDuplicateProperties.mergeAccessors(gltfWithExtras);
+    removeDuplicatePrimitives(gltfWithExtras);
+    combinePrimitives(gltfWithExtras);
+    combineMeshes(gltfWithExtras);
```
Comment (lilleyse) [inline lib/Pipeline.js]: I think we talked offline about this a while ago, but I forget the conversation...

If possible we should try to avoid the second cleanup, and ideally our stages shouldn't be producing unoptimized gltf's. Sorry if I forgot the reason behind this.

File: lib/PrimitiveHelpers.js [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+        if (indices.indexOf(index) >= 0) {
+            return true;
+        }
+        accessorReader.next();
+    }
+    return false;
+}
+
+function transformPrimitives(gltf, primitives, transform) {
+    var inverseTranspose = new Matrix4();
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: Can't you just do a simple loop over the position and normal arrays, transforming each set of 3 values with the transform and then writing back into the buffer. I don't see the need for involving indices.

File: lib/PrimitiveHelpers.js [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+            primitives = primitives.concat(mesh.primitives);
+        }
+    }
+    return primitives;
+}
+
+function generateIndices(v, k) {
+    void(v); // v is an unused parameter
+    return k;
+}
```
Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: If still needed, just put this code in `transformPrimitives`.

File: lib/combineNodes.js [hunk 15]
```diff
@@ -265,123 +261,12 @@ function meshHasConflict(gltf, mesh, compareMesh) {
         var primitive = primitives[i];
         for (var j = 0; j < comparePrimitivesLength; j++) {
             var comparePrimitive = comparePrimitives[j];
-            if (primitivesShareAttributeAccessor(primitive, comparePrimitive)) {
-                if (primitivesHaveOverlappingIndexAccessors(gltf, primitive, comparePrimitive)) {
+            if (PrimitiveHelpers.primitivesShareAttributeAccessor(primitive, comparePrimitive)) {
+                if (PrimitiveHelpers.primitivesHaveOverlappingIndexAccessors(gltf, primitive, comparePrimitive)) {
```
Comment (lilleyse) [inline lib/combineNodes.js]: Create a function for this in `PrimitiveHelpers` that is called here and inside `markPrimitiveConflicts`.

File: lib/combinePrimitives.js [hunk 16]
```diff
@@ -1,255 +1,252 @@
...
-    if (defined(gltf.accessors) && defined(gltf.bufferViews) && defined(gltf.buffers)) {
-        var meshes = gltf.meshes;
-        if (defined(meshes)) {
-            for (var meshId in meshes) {
-                if (meshes.hasOwnProperty(meshId)) {
-                    var mesh = meshes[meshId];
-                    var primitives = mesh.primitives;
-                    if (defined(primitives)) {
-                        mesh.primitives = combineMeshPrimitives(gltf, meshId, primitives);
+    PrimitiveHelpers.markPrimitiveConflicts(gltf, PrimitiveHelpers.getAllPrimitives(gltf));
```
Comment (lilleyse) [inline lib/combinePrimitives.js]: It's a little awkward to save the conflicts in the pipeline object. If possible try to keep it in here.

**Comment by lilleyse** [comment]
Thanks for the update @lasalvavida 
It's been a while since I looked at this so fill me in case my comments don't make sense. I'll do some testing tomorrow.

**Comment by lilleyse** [comment]
I'm having issues converting this model: 
[dragon_low.gltf.txt](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/files/567387/dragon_low.gltf.txt)

File: lib/PrimitiveHelpers.js (hunk 13 above)
Comment (lasalvavida) [inline lib/PrimitiveHelpers.js]: However, this should use `AccessorReader` instead of `readAccessor`. I will make that change.

Comment (lilleyse) [inline lib/PrimitiveHelpers.js]: Ah right, thanks for the correction.

**Comment by lasalvavida** [comment]
This is ready for review. Referring to my original benchmarks the stats are now:

|  | Original | Master | Master w/ combine primitives |
| --- | --- | --- | --- |
| File size | 10,414 KB | 6,091 KB | 6,023 KB |
| Nodes | 208 | 179 | 179 |
| Meshes | 109 | 163 | 61 |
| Draw Calls | 236 | 273 | 196 |

**Comment by lilleyse** [comment]
This is really close. Every model I've tested works fine. Those numbers are nice!

File: lib/cesiumGeometryToGltfPrimitive.js [hunk 17]
```diff
@@ -3,81 +3,22 @@ var Cesium = require('cesium');
...
-    var attributes = primitive.attributes;
-    for (var attributeSemantic in attributes) {
-        if (attributes.hasOwnProperty(attributeSemantic)) {
-            if (attributeSemantic.indexOf(semantic) === 0) {
-                return attributeSemantic;
-            }
-        }
+    var semantics = getPrimitiveAttributeSemantics(primitive, semantic);
+    if (semantics.length <= 0) {
+        primitive.attributes[semantic] = createAccessor(gltf, packedLength, 'VEC3', WebGLConstants.FLOAT, WebGLConstants.ARRAY_BUFFER);
```
Comment (lasalvavida) [inline lib/cesiumGeometryToGltfPrimitive.js]: Ah yes, you're right

**Comment by lasalvavida** [comment]
Updated

**Comment by lilleyse** [comment]
Nice, I'm glad to finally get this in! Also because of adjustments here the pipeline just runs a lot faster.
